packages/svelte/src/internal/client/reactivity/batch.js

```diff
@@ -601,15 +601,35 @@ function flush_queued_effects(effects) {
 
 			// If update_effect() has a flushSync() in it, we may have flushed another flush_queued_effects(),
 			// which already handled this logic and did set eager_block_effects to null.
-			if (eager_block_effects?.length > 0) {
-				// TODO this feels incorrect! it gets the tests passing
+			if (eager_block_effects?.size > 0) {
 				old_values.clear();
 
 				for (const e of eager_block_effects) {
-					update_effect(e);
+					// Skip eager effects that have already been unmounted
+					if ((e.f & (DESTROYED | INERT)) !== 0) continue;
+
+					// Run effects in order from ancestor to descendant, else we could run into nullpointers
+					/** @type {Effect[]} */
+					const ordered_effects = [];
+					let ancestor = e.parent;
+					while (ancestor !== null) {
+						if (eager_block_effects.has(ancestor)) {
+							eager_block_effects.delete(ancestor);
+							ordered_effects.unshift(ancestor);
+						}
+						ancestor = ancestor.parent;
+					}
+
+					ordered_effects.push(e);
+
+					for (const e of ordered_effects) {
+						// Skip eager effects that have already been unmounted
+						if ((e.f & (DESTROYED | INERT)) !== 0) continue;
```

Comment:
Wouldn't it make more sense to make this check in the while loop so that the effect doesn't get added to `ordered_effects`?

Reply:
I think the idea might be that an effect might be destroyed by an earlier parent effect?